packages/dds/tree/src/brand.ts

```diff
@@ -0,0 +1,108 @@
+ * - allow use as {@link Opaque} branded type (not assignable to `ValueType`, but captures `ValueType`).
+ *
+ * See {@link MakeNominal} for some more details.
+ *
+ * Do not use this class with `instanceof`: this will always be false at runtime,
+ * but the compiler may think its true in some cases.
+ */
+export abstract class BrandedType<ValueType, Name extends string> {
+    protected _typeCheck?: Invariant<ValueType>;
+    /**
+     * Compile time only marker to make type checking more strict.
+     * This field will not exist at runtime and accessing it is invalid.
+     * See {@link Brand} for details.
+     */
+    protected readonly _type_brand!: Name;
+
+    /**
+     * This class should never exist at runtime, so make it un-constructable.
+     */
+    private constructor() {}
+}
+
+/**
+ * Converts a Branded type into an "opaque" handle.
+ * This prevents the value from being used directly, but does not fully type erase it
+ * (and this its not really fully opaque):
+ * The type can be recovered using {@link extractFromOpaque},
+ * however if we assume only code that produces these "opaque" handles does that conversion,
+ * they can function like opaque handles.
+ */
+export type Opaque<T extends Brand<any, string>> = T extends Brand<
+    infer ValueType,
+    infer Name
+>
+    ? BrandedType<ValueType, Name>
+    : never;
+
+/**
+ * See {@link extractFromOpaque}.
+ */
+export type ExtractFromOpaque<TOpaque extends BrandedType<any, string>> =
+    TOpaque extends BrandedType<infer ValueType, infer Name>? Brand<ValueType, Name> : never;
+
+type ValueFromBranded<T extends BrandedType<any, string>> =
+    T extends BrandedType<infer ValueType, string> ? ValueType : never;
+type NameFromBranded<T extends BrandedType<any, string>> = T extends BrandedType< any, infer Name> ? Name : never;
+
+/**
+ * Converts a {@link Opaque} handle to the underlying branded type.
+ *
+ * It is assumed that only code that produces these "opaque" handles does this conversion,
+ * allowing these handles to be considered opaque.
+ */
+export function extractFromOpaque<TOpaque extends BrandedType<any, string>>(
+    value: TOpaque,
+): ExtractFromOpaque<TOpaque> {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return value as ExtractFromOpaque<TOpaque>;
+}
+
+/**
+ * Adds a type {@link Brand} to a value.
+ *
+ * Only do this when specifically allowed by the requirements of the type being converted to.
+ */
+export function asBranded<T extends Brand<any, string>>(
+    value: T extends BrandedType<infer ValueType, string> ? ValueType : never,
+): T {
+    return value as T;
+}
+
+/**
+ * Adds a type {@link Brand} to a value, returning it as a  {@link Opaque} handle.
+ *
+ * Only do this when specifically allowed by the requirements of the type being converted to.
+ */
+ export function asOpaque<T extends BrandedType<any, string>>(
+    value: ValueFromBranded<T>,
```

Comment:
Do we want to put restrictions on `value` here like we do for the `value` parameter in `asBranded`? I.e. `never` if `T` is not a `BrandedType`.

Reply:
Its unclear to me what you want changed here.

T is BrandedType: this is enforced through the extends constraint here. It is enforced again in ValueFromBranded, which also (due to language limitations) handles the case where T is not BrandedType (which is impossible) by returning never.

Reply:
If I write this:
`const b = asBranded(3);`
I get a compile error. But if I write this:
`const o = asOpaque(3); // BrandedType<any, string>`
I get a `BrandedType<any, string>` which is useless (or at least as useless as a `Brand<any, string>`). I'm hoping we can make this case emit a compile error as well.

Reply:
I allow `brand` to infer its types from the output type. This seems relatively safe, as usually when you brand something you are doing so at a location where you need the type brand, and you know what brand is needed. If in doute you can always specify it explicitly.

brandOpaque on the other hand is intended to remove type information. If I make it work like `brand` (return `T`), when given a branded (non-opaque) type, it will produce a branded non-opaque type, failing to remove any type information like expected. We specifically want  to avoid having toOpaque ever return some type that extends BrandedType (like a non opaque branded type): instead it must return exactly a BrandedType. This results in different inference requiring most usages to provide type explicitly.

The fact that it happens to compile and return a useless value is just a side effect of the inference, and is just about as good as failing to build (because who ever uses the value will fail to build).

That said, an untyped opaquer handle can leak an `any` value via `extractFromOpaque`, which is bad, so I fixed that.

We might be abler to get brandOpaque to infer `never` like `brand` , but I think that would require making ValueType covariant instead of invariant which is unsound if anyone uses a mutable type for it. Thus I instead just added an explicit check for any. Kinda ugly, but seems to do what we want.

I think it should make you happy now.

Reply:
Thanks for the thoughtful reply - I agree it's a little weird but still better!